### PR TITLE
feat(tier4_perception_rviz_plugin): interactive pedestrian for rviz plugin

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Related Issue(required)

- https://github.com/autowarefoundation/autoware.universe/issues/533

## Description(required)

Add interactive dummy pedestrian plugin tool that can be manipulated by using the mouse in order to simulate Ego's behavior in crosswalks more realistically and easily in Psim.

https://user-images.githubusercontent.com/44889564/158741228-69a0a349-79d1-4b70-a595-ceaa47644043.mp4

## Review Procedure(required)

please see the above video.

key commands are...

| action | key command |
| ---- | ---- |
| ADD  | Shift + Click Right Button |
| MOVE | Hold down Right Button + Drug and Drop |
| DELETE | Alt + Click Right Button |

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [x] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: common/tier4_perception_rviz_plugin/src/tools/interactive_pedestrian.cpp [hunk 1]
```diff
@@ -0,0 +1,377 @@
+// Copyright 2020 Tier IV, Inc.
```
Comment (yukkysaito) [inline common/tier4_perception_rviz_plugin/src/tools/interactive_pedestrian.cpp:1]:
```suggestion
// Copyright 2022 Tier IV, Inc.
```

**Comment by yukkysaito** [comment]
I think you can extend the existing pedestrian plugin. then the tool property has an interaction mode checkbox.

**Comment by satoshi-ota** [comment]
> @satoshi-ota I think you can extend the existing pedestrian plugin. then the tool property has an interaction mode checkbox.

Thank you for your advise :+1: 
Consider that ways to implement interactive dummy pedestrian by extending the existing plugins.

File: common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp [hunk 2]
```diff
@@ -47,25 +47,194 @@
...
+  }
+
+  tf_map2object = tf_map2object_origin * tf_object_origin2moved_object;
+}
+
+void InteractivePedestrian::update(const Ogre::Vector3 & point)
+{
+  velocity_ = point - point_;
+  point_ = point;
+  theta_ = velocity_.length() < std::numeric_limits<double>::epsilon()
```
Comment (isamu-takagi) [inline common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp:113]: I don't recommend using machine epsilon for floating value comparisons. Generally, the error of the calculation result will be larger.

In this case, the point has three axes, so if each has an epsilon error but represents the same position, the length of velocity will be root 3 epsilon.

Moreover, length uses xyz, but atan2 does xy only. So atan2(0, 0) is called if velocity_ is (0, 0, epsilon)

File: common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp [hunk 3]
```diff
@@ -165,6 +340,108 @@ void PedestrianInitialPoseTool::onPoseSet(double x, double y, double theta)
...
+    const auto point = get_point_from_mouse(event);
+    if (point) {
+      if (event.shift()) {
+        const auto uuid = objects_.create(point.value());
+        publishObjectMsg(uuid.get(), Object::ADD);
+      } else if (event.alt()) {
+        const auto uuid = objects_.remove(point.value());
+        publishObjectMsg(uuid.get(), Object::DELETE);
+      } else {
+        const auto uuid = objects_.select(point.value());
```
Comment (isamu-takagi) [inline common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp:417]: This uuid is not used.

File: common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp (hunk 2 above)
Comment (satoshi-ota) [reply common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp:113]: OK, I see :+1: In short, the following implementation is better than original one, right?

```
theta_ = (velocity_.x < 1e-3 && velocity_.y < 1e-3) ? theta_ : atan2(velocity_.y, velocity_.x);
```

Comment (isamu-takagi) [reply common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp:113]: I think so. (But I don't know what value is suitable for epsilon. It's very difficult.)

Comment (satoshi-ota) [reply common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp]: For now, I'll fix as above example.

https://github.com/autowarefoundation/autoware.universe/pull/534/commits/133f3f7726711f5ef8222d6b18186011deefa3f7

File: common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp (hunk 3 above)
Comment (satoshi-ota) [reply common/tier4_perception_rviz_plugin/src/tools/pedestrian_pose.cpp]: fix in https://github.com/autowarefoundation/autoware.universe/pull/534/commits/133f3f7726711f5ef8222d6b18186011deefa3f7

**Comment by isamu-takagi** [comment]
LGTM, but it needs to be reviewed by others since I wrote the original code.

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**